abi3 fallback for older interpreters incorrectly sets Py_LIMITED_API in pyo3 config

## Bug Description

When maturin detects abi3 support and an interpreter older than the abi3 minimum is requested (e.g., Python 3.10 when abi3 targets ≥ 3.11), it falls back to building a version-specific wheel. However, the generated `pyo3-config-*.txt` for the fallback build still appears to have `abi3=true` set, which causes `Py_LIMITED_API` to be defined during compilation.

This breaks crates that use pyo3 APIs gated behind `cfg(any(not(Py_LIMITED_API), Py_3_XX))` — for example, `pyo3::buffer` which requires either non-limited-API or Python ≥ 3.11.

**This is a regression introduced in maturin 1.13.x** (likely related to #3126 "fall back to placeholder for abi3 when found interpreters are too old"). The same project builds successfully with maturin 1.12.6.

## Reproduction

Project: https://github.com/trim21/bencode-rs/pull/240

### pyproject.toml (relevant sections)

```toml
[build-system]
requires = ["maturin==1.13.2"]
build-backend = "maturin"

[tool.maturin]
features = [
  { feature = "pyo3/abi3-py311", python-version = ">=3.11" },
]
```

### Cargo.toml (relevant sections)

```toml
[dependencies]
pyo3 = { version = "0.28.3", features = ['generate-import-lib'] }
```

### Build command

```
maturin build --release --out dist --interpreter '3.11' --interpreter '3.10'
```

### Steps to reproduce

1. Use a pyo3 project that imports `pyo3::buffer` (which is gated behind `cfg(any(not(Py_LIMITED_API), Py_3_11))`)
2. Configure conditional abi3 features in `pyproject.toml`: `{ feature = "pyo3/abi3-py311", python-version = ">=3.11" }`
3. Build with `--interpreter '3.11' --interpreter '3.10'`

### Expected behavior

- First build: abi3 wheel for Python ≥ 3.11 ✅ (works)
- Second build: version-specific wheel for Python 3.10 with `abi3=false` in pyo3 config ✅

### Actual behavior

- First build: abi3 wheel for Python ≥ 3.11 ✅ (works)
- Second build: version-specific wheel for Python 3.10, but the generated pyo3 config appears to still have `abi3=true`, causing `Py_LIMITED_API` to be defined ❌

## Error output

```
📦 Built wheel for abi3 Python ≥ 3.11 to dist/bencode_rs-0.0.9-cp311-abi3-manylinux_2_17_x86_64.manylinux2014_x86_64.whl
⚠️ Warning: abi3 does not yet support CPython 3.10 so the build artifacts will be version-specific.
error[E0432]: unresolved import `pyo3::buffer`
  --> src/encode.rs:2:11
  |
2 | use pyo3::buffer::{PyBuffer, ReadOnlyCell};
  |           ^^^^^^^ could not find `buffer` in `pyo3`
  |
note: found an item that was configured out
  --> pyo3-0.28.3/src/lib.rs:413:9
    |
413 | pub mod buffer;
    |         ^^^^^^
    |
::: pyo3-0.28.3/src/buffer.rs:1:11
    |
  1 | #![cfg(any(not(Py_LIMITED_API), Py_3_11))]
    |           ------------------------------ the item is gated here
```

## Environment

- maturin version: 1.13.2 (regression from 1.12.6)
- pyo3 version: 0.28.3
- Python interpreters: 3.10, 3.11
- Platforms affected: all (Linux x86_64, Linux aarch64, macOS arm64, etc.)

## CI Run

Full CI failure logs: https://github.com/trim21/bencode-rs/actions/runs/25625690761

--- a/src/bridge/mod.rs
+++ b/src/bridge/mod.rs
@@ -9,6 +9,7 @@ use serde::Deserialize;
 
 use crate::python_interpreter::{
     MAXIMUM_PYPY_MINOR, MAXIMUM_PYTHON_MINOR, MINIMUM_PYPY_MINOR, MINIMUM_PYTHON_MINOR,
+    PythonInterpreter,
 };
 
 /// pyo3 binding crate
@@ -335,16 +336,35 @@ impl BridgeModel {
         }
     }
 
-    /// Is using abi3
-    pub fn is_abi3(&self) -> bool {
+    /// Returns `true` if the bridge model carries stable-ABI metadata (e.g. abi3).
+    ///
+    /// This is a project-level check — it does not consider whether a particular
+    /// interpreter meets the abi3 minimum version.  For per‑interpreter checks
+    /// use [`is_abi3_for_interpreter`](Self::is_abi3_for_interpreter).
+    pub fn has_stable_abi(&self) -> bool {
         self.pyo3()
-            .and_then(|pyo3| match pyo3.stable_abi {
-                Some(stable_abi) => match stable_abi.kind {
-                    StableAbiKind::Abi3 => Some(true),
-                },
-                None => None,
+            .and_then(|pyo3| pyo3.stable_abi.as_ref())
+            .is_some()
+    }
+
+    /// Check whether abi3 should be enabled for a specific interpreter.
+    ///
+    /// Returns `true` only when the bridge model has stable abi support **and**
+    /// the given interpreter meets the abi3 minimum version.  Version‑specific
+    /// fallback builds (e.g. Python 3.10 when abi3 targets ≥ 3.11) return
+    /// `false` so that `Py_LIMITED_API` is not defined and interpreter‑specific
+    /// linker names are used.
+    pub fn is_abi3_for_interpreter(&self, interpreter: &PythonInterpreter) -> bool {
+        self.pyo3()
+            .and_then(|pyo3| pyo3.stable_abi.as_ref())
+            .is_some_and(|stable_abi| {
+                match stable_abi.version.min_version() {
+                    Some((major, minor)) => {
+                        (interpreter.major as u8, interpreter.minor as u8) >= (major, minor)
+                    }
+                    None => true, // CurrentPython → always compatible
+                }
             })
-            .is_some_and(|x| x)
     }
 
     /// free-threaded Python support

--- a/src/bridge/mod.rs
+++ b/src/bridge/mod.rs
@@ -350,20 +350,22 @@ impl BridgeModel {
     /// Check whether abi3 should be enabled for a specific interpreter.
     ///
     /// Returns `true` only when the bridge model has stable abi support **and**
-    /// the given interpreter meets the abi3 minimum version.  Version‑specific
-    /// fallback builds (e.g. Python 3.10 when abi3 targets ≥ 3.11) return
-    /// `false` so that `Py_LIMITED_API` is not defined and interpreter‑specific
-    /// linker names are used.
+    /// the given interpreter supports the stable ABI **and** meets the abi3
+    /// minimum version. Version‑specific fallback builds (e.g. Python 3.10 when
+    /// abi3 targets ≥ 3.11) return `false` so that `Py_LIMITED_API` is not
+    /// defined and interpreter‑specific linker names are used.
     pub fn is_abi3_for_interpreter(&self, interpreter: &PythonInterpreter) -> bool {
+        if !interpreter.has_stable_api() {
+            return false;
+        }
+
         self.pyo3()
             .and_then(|pyo3| pyo3.stable_abi.as_ref())
-            .is_some_and(|stable_abi| {
-                match stable_abi.version.min_version() {
-                    Some((major, minor)) => {
-                        (interpreter.major as u8, interpreter.minor as u8) >= (major, minor)
-                    }
-                    None => true, // CurrentPython → always compatible
+            .is_some_and(|stable_abi| match stable_abi.version.min_version() {
+                Some((major, minor)) => {
+                    (interpreter.major as u8, interpreter.minor as u8) >= (major, minor)
                 }
+                None => true, // CurrentPython → compatible when stable ABI is supported
             })
     }
 

--- a/src/build_options.rs
+++ b/src/build_options.rs
@@ -307,7 +307,7 @@ mod tests {
         // find_bridge alone never includes conditional features → no abi3
         let bridge = find_bridge(&metadata, None).unwrap();
         assert!(
-            !bridge.is_abi3(),
+            !bridge.has_stable_abi(),
             "find_bridge should not infer abi3 from conditional features"
         );
 
@@ -328,7 +328,10 @@ mod tests {
             std::slice::from_ref(&py310),
         )
         .unwrap();
-        assert!(!bridge.is_abi3(), "should not infer abi3 for Python 3.10");
+        assert!(
+            !bridge.has_stable_abi(),
+            "should not infer abi3 for Python 3.10"
+        );
 
         // With a Python 3.11 interpreter, condition matches → abi3
         let py311 = crate::PythonInterpreter::from_config(
@@ -348,7 +351,7 @@ mod tests {
             std::slice::from_ref(&py311),
         )
         .unwrap();
-        assert!(bridge.is_abi3(), "should infer abi3 for Python 3.11");
+        assert!(bridge.has_stable_abi(), "should infer abi3 for Python 3.11");
 
         // With mixed interpreters [3.10, 3.11], abi3 IS inferred because
         // at least one interpreter (3.11) matches the condition. This is safe
@@ -358,7 +361,7 @@ mod tests {
         let bridge =
             upgrade_bridge_abi3(base_bridge, &metadata, Some(&pyproject), &[py310, py311]).unwrap();
         assert!(
-            bridge.is_abi3(),
+            bridge.has_stable_abi(),
             "should infer abi3 for mixed [3.10, 3.11] (build_stable_abi_wheels handles the split)"
         );
     }

--- a/src/ci/github/render.rs
+++ b/src/ci/github/render.rs
@@ -87,7 +87,7 @@ pub(crate) fn generate_github(
     sdist: bool,
     min_python_minor: Option<u8>,
 ) -> Result<String> {
-    let is_abi3 = bridge_model.is_abi3();
+    let is_abi3 = bridge_model.has_stable_abi();
     let is_bin = bridge_model.is_bin();
     let setup_python = resolved.pytest
         || matches!(

--- a/src/compile.rs
+++ b/src/compile.rs
@@ -620,7 +620,9 @@ fn configure_macos_pyo3_linker_args(
     }
 
     // install_name is specific to the top-level output, so keep it in cargo_rustc.args
-    let so_filename = if bridge_model.is_abi3() {
+    let use_abi3_suffix =
+        python_interpreter.is_some_and(|i| bridge_model.is_abi3_for_interpreter(i));
+    let so_filename = if use_abi3_suffix {
         format!("{module_name}.abi3.so")
     } else {
         python_interpreter
@@ -852,7 +854,10 @@ fn configure_pyo3_env(
     target: &Target,
     target_triple: &str,
 ) -> Result<()> {
-    if bridge_model.is_abi3() {
+    // Only set PYO3_NO_PYTHON for true abi3 builds where the interpreter meets
+    // the abi3 minimum version.  Version-specific fallback builds (e.g. Python
+    // 3.10 when abi3 targets ≥ 3.11) must not set PYO3_NO_PYTHON.
+    if python_interpreter.is_some_and(|i| bridge_model.is_abi3_for_interpreter(i)) {
         let is_pypy_or_graalpy = python_interpreter
             .map(|p| p.interpreter_kind.is_pypy() || p.interpreter_kind.is_graalpy())
             .unwrap_or(false);
@@ -892,7 +897,8 @@ fn configure_pyo3_env(
             // and legacy pyo3 versions
             build_command.env("PYTHON_SYS_EXECUTABLE", &interpreter.executable);
         } else if bridge_model.is_pyo3() && env::var_os("PYO3_CONFIG_FILE").is_none() {
-            let pyo3_config = interpreter.pyo3_config_file(target, bridge_model.is_abi3());
+            let use_abi3 = bridge_model.is_abi3_for_interpreter(interpreter);
+            let pyo3_config = interpreter.pyo3_config_file(target, use_abi3);
             let maturin_target_dir = ensure_target_maturin_dir(&context.project.target_dir);
             let config_file = maturin_target_dir.join(format!(
                 "pyo3-config-{}-{}.{}.txt",

